### PR TITLE
Added Controllers and multithreading support for some

### PR DESCRIPTION
**Pull request by Symbyos**
Feel free to tell me if it follows the repo guidelines.

### CODE REVIEW

File: AutoFinance.Broker/InteractiveBrokers/Constants/TwsHistoricalDataRequestType.cs [hunk 1]
```diff
@@ -27,6 +27,11 @@ public class TwsHistoricalDataRequestType : ITwsStringParameter
         /// </summary>
         private static TwsHistoricalDataRequestType trades = new TwsHistoricalDataRequestType("TRADES");
 
+        /// <summary>
+        /// Implied volatility
+        /// </summary>
+        private static TwsHistoricalDataRequestType iv = new TwsHistoricalDataRequestType("OPTION_IMPLIED_VOLATILITY");
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Constants/TwsHistoricalDataRequestType.cs:33]:
```suggestion
                private static TwsHistoricalDataRequestType impliedVolatility = new TwsHistoricalDataRequestType("OPTION_IMPLIED_VOLATILITY");
```

File: AutoFinance.Broker/InteractiveBrokers/Constants/TwsHistoricalDataRequestType.cs [hunk 2]
```diff
@@ -61,6 +66,11 @@ private TwsHistoricalDataRequestType(string type)
         /// </summary>
         public static TwsHistoricalDataRequestType Trades { get => trades; }
 
+        /// <summary>
+        /// Gets trade
+        /// </summary>
+        public static TwsHistoricalDataRequestType Iv { get => iv; }
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Constants/TwsHistoricalDataRequestType.cs:72]:
```suggestion
        public static TwsHistoricalDataRequestType ImpliedVolatility { get => impliedVolatility; }
```

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 3]
```diff
@@ -696,58 +834,148 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
+        /// Set the type for the market data feed
         /// </summary>
-        /// <param name="underlyingSymbol">The underlying symbol</param>
-        /// <param name="exchange">The exchange</param>
-        /// <param name="underlyingSecType">The underlying security type</param>
-        /// <param name="underlyingConId">The underlying contract ID, retrieved from the Contract Details call</param>
+        /// <param name="marketDataTypeId">The feed level</param>
+        public void RequestMarketDataType(int marketDataTypeId)
+        {
+            this.clientSocket.RequestMarketDataType(marketDataTypeId);
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:854]: Calling this method will produce a callback on the `MarketDataTypeEvent` of the `TwsCallbackHandler`.

Following the pattern, this method should be an async method which listens for the event before returning. `GetAccountDetailsAsync` is a good template method to follow.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 4]
```diff
@@ -696,58 +834,148 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
+        /// <param name="accountCode">The account code</param>
+        /// <param name="modelCode">The model code</param>
+        /// <returns>The PnL account update event from TWS.</returns>
+        public Task<PnLEventArgs> RequestPnL(
+            string accountCode,
+            string modelCode)
+        {
+            // Set the operation to cancel after 5 seconds
+            CancellationTokenSource tokenSource = new CancellationTokenSource(5000);
+            return this.RequestPnL(accountCode, modelCode, tokenSource.Token);
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:869]: `RequestPnl` is a subscription method, so there's no need to cancel with a token. `RequestPnlSingle`, however, returns the PNL without subscribing, so should listen for the callback (with a cancellation token), and return when the event is fired.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 5]
```diff
@@ -696,58 +834,148 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
+            return taskSource.Task;
+        }
+
+        /// <summary>
+        /// Request PnL update cancelation
+        /// </summary>
+        /// <param name="reqId">The request Id</param>
+        public void CancelPnL(int reqId)
+        {
+            this.clientSocket.CancelPnL(reqId);
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:913]: use `this.twsRequestIdGenerator.GetNextRequestId()` to generate the `reqId` and remove it from the parameter

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 6]
```diff
@@ -696,58 +834,148 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
             return taskSource.Task;
         }
 
+        /// <summary>
+        /// Request PnL update cancelation
+        /// </summary>
+        /// <param name="reqId">The request Id</param>
+        public void CancelPnLSingle(int reqId)
+        {
+            this.clientSocket.CancelPnLSingle(reqId);
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:976]: use `this.twsRequestIdGenerator.GetNextRequestId()` to generate the `reqId` and remove it from the parameter

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 5 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:913]: Following the documentation https://interactivebrokers.github.io/tws-api/pnl.html#portfolio_window_pnl it seems we need to pass the same reqId as the one provided when requesting the PnL.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 6 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:976]: Following the documentation https://interactivebrokers.github.io/tws-api/pnl.html#portfolio_window_pnl it seems we need to pass the same reqId as the one provided when requesting the PnL.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 7]
```diff
@@ -696,58 +834,148 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
         {
-            var taskSource = new TaskCompletionSource<List<SecurityDefinitionOptionParameterEventArgs>>();
+            PnLEventArgs pnlUpdateEvent = null;
+            var taskSource = new TaskCompletionSource<PnLEventArgs>();
+            EventHandler<PnLEventArgs> pnlUpdateEventHandler = null;
 
-            List<SecurityDefinitionOptionParameterEventArgs> securityDefinitionEvents = new List<SecurityDefinitionOptionParameterEventArgs>();
+            pnlUpdateEventHandler = (sender, args) =>
+            {
+                pnlUpdateEvent = args;
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:890]: https://interactivebrokers.github.io/tws-api/pnl.html

Looks like this is a subscription method, so you don't want to return the PNL event immediately when it's fired.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/ITwsControllerBase.cs [hunk 8]
```diff
@@ -173,5 +179,45 @@ public interface ITwsControllerBase
             string exchange,
             string underlyingSecType,
             int underlyingConId);
+
+        /// <summary>
+        /// Set the type for the market data feed
+        /// </summary>
+        /// <param name="marketDataTypeId">The feed level</param>
+        void RequestMarketDataType(int marketDataTypeId);
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/ITwsControllerBase.cs:187]: Add an integration test for at least `RequestMarketDataType` to try to keep our implemented API surface covered.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 3 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:854]: Ok.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 4 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:869]: Ok.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 7 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs]: Ok.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 3 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs]: It seems that `MarketDataTypeEvent` is triggered by `reqMktData` and not `reqMarketDataType`.
This event has for second argument `int reqId` which corresponds to `int tickerId` from `reqMktData`. 
As seen in https://interactivebrokers.github.io/tws-api/interfaceIBApi_1_1EWrapper.html#a79918660257f0e7ce77309bf3bccf13c.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 9]
```diff
@@ -849,55 +858,60 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
         /// </summary>
-        /// <param name="accountCode">The account code</param>
-        /// <param name="modelCode">The model code</param>
-        /// <returns>The PnL account update event from TWS.</returns>
-        public Task<PnLEventArgs> RequestPnL(
-            string accountCode,
-            string modelCode)
+        /// <param name="marketDataTypeId">The feed level</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public Task RequestMarketDataTypeAsync(int marketDataTypeId, CancellationToken cancellationToken)
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:874]: Make an integration test around this method

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 10]
```diff
@@ -849,55 +858,60 @@ public Task<List<PositionStatusEventArgs>> RequestPositions(CancellationToken ca
...
         /// </summary>
         /// <param name="accountCode">The account code</param>
         /// <param name="modelCode">The model code</param>
-        /// <param name="cancellationToken">The cancellation token used to cancel the request</param>
         /// <returns>The PnL account update event from TWS.</returns>
         public Task<PnLEventArgs> RequestPnL(
             string accountCode,
-            string modelCode,
-            CancellationToken cancellationToken)
+            string modelCode)
```
Comment (amittleider) [inline AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:902]: Make an integration test around demonstrating how to use it (requestPnl + cancelPnl, requestPnlSingle, cancelPnlSingle)

Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:1248]: Done.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 9 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs]: Done.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 3 above)
Comment (Symbyos) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs]: I confirm my previous comment as you will see in the integration tests.

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs [hunk 11]
```diff
@@ -98,21 +94,541 @@ public TwsControllerBase(ITwsClientSocket clientSocket, ITwsCallbackHandler twsC
...
+        /// <param name="twsCallbackHandler">The callback handler</param>
+        /// <param name="twsClientSocket">The client socket</param>
+        /// <param name="cancellationToken">The cancellation token used to cancel the request</param>
+        /// <returns>The security definitions for options</returns>
+        public static Task<List<SecurityDefinitionOptionParameterEventArgs>> RequestSecurityDefinitionOptionParameters(
+            string underlyingSymbol,
+            string exchange,
+            string underlyingSecType,
+            int underlyingConId,
+            ITwsRequestIdGenerator twsRequestIdGenerator,
```
[the remark this answers is not in the document]
Comment (popojargo) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:120]: He made the function a static helper...

**Comment by amittleider** [comment]
Why though?

Le jeu. 11 juin 2020 à 02:07, Alexis Côté <notifications@github.com> a
écrit :

File: AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (amittleider) [reply AutoFinance.Broker/InteractiveBrokers/Controllers/TwsControllerBase.cs:120]: I still wonder why?

**Comment by amittleider** [comment]
Thanks for contributing. Your effort is appreciated.

**Pull request merged**